### PR TITLE
NavModel2

### PR DESCRIPTION
**Pull request by zsoltk**
## Description

Adds:

- POC for new `NavModel`
- Converted versions of `BackStack`, `Spotlight`, and `Promoter`
- Some experiments to test it all out standalone in `sandbox2`

It's missing at this point almost all of the `1.x` under the hood mechanics:

- Not connected to `Node` / `ParentNode`
- No child `Node` resolution
- No child `Node` removal
- `OnScreenResolver` not taken into account
- etc.
 
Many pieces of the code are copy-pasted from `navigation` → `navigation2`, `navmodel` → `navmodel2` packages to make it easier for a gradual change instead of breaking everything all at once.

Contains TODOs and many things to be improved for sure. The PR is against `2.x`, won't affect any code we release now. I'll merge it if it doesn't break CI so that we can collaborate on it quicker, but nonetheless feel free to collect any feedback on this ticket even after it is merged, and we can come back to addressing them.

...

## Check list

- [ ] I have updated `CHANGELOG.md` if required.
- [ ] I have updated documentation if required.

### CODE REVIEW

File: libraries/core/build.gradle.kts [hunk 1]
```diff
@@ -37,6 +37,9 @@ dependencies {
     api(libs.androidx.lifecycle.common)
     api(libs.compose.animation.core)
     api(libs.compose.runtime)
+    api(libs.compose.material)
+    api(libs.compose.ui.tooling)
+    api(libs.compose.ui.ui)
```
Comment (zsoltk) [inline libraries/core/build.gradle.kts]: This is a shortcut for not extracting transitions-related code for now. It def shouldn't live in core in the long run.

File: build.gradle.kts [hunk 2]
```diff
@@ -17,6 +17,8 @@ plugins {
     id("com.autonomousapps.dependency-analysis") version libs.versions.dependencyAnalysis.get()
     id("release-dependencies-diff-compare")
     id("release-dependencies-diff-create") apply false
+    id("com.android.application") version "7.2.2" apply false
```
Comment (LachlanMcKee) [inline build.gradle.kts:20]: Why were these changes necessary?

File: libraries/core/build.gradle.kts [hunk 3]
```diff
@@ -37,6 +37,9 @@ dependencies {
     api(libs.androidx.lifecycle.common)
     api(libs.compose.animation.core)
     api(libs.compose.runtime)
+    api(libs.compose.foundation)
```
Comment (LachlanMcKee) [inline libraries/core/build.gradle.kts:40]: If these aren't going to required in the future, maybe add a TODO saying to remove them before merging into main or whatever the branch is going forward.

**Review by LachlanMcKee — COMMENTED** [review]
Half of this PR is unrelated (because the 2.x branch was not up to date), so it's very difficult to review this in isolation. Any tips how to review (post-merge?)

**Pull request merged**